src/cc/frontends/clang/kbuild_helper.cc

```diff
@@ -107,4 +113,81 @@ int KBuildHelper::get_flags(const char *uname_machine, vector<string> *cflags) {
   return 0;
 }
 
+static inline int file_exists(const char *f)
+{
+  struct stat buffer;
+  return (stat(f, &buffer) == 0);
+}
+
+static inline int proc_kheaders_exists(void)
+{
+  return file_exists(PROC_KHEADERS_PATH);
+}
+
+static inline int extract_kheaders(std::string &dirpath, struct utsname &uname_data)
+{
+  char tar_cmd[256], dirpath_tmp[256];
+  int ret;
+  bool module = false;
+
+  if (!proc_kheaders_exists()) {
+    ret = system("modprobe kheaders");
+    if (ret)
+      return ret;
+    module = true;
+    if (!proc_kheaders_exists()) {
+      ret = -1;
+      goto cleanup;
+    }
+  }
+
+  snprintf(dirpath_tmp, 256, "/tmp/kheaders-%s-XXXXXX", uname_data.release);
+  if (mkdtemp(dirpath_tmp) == NULL) {
+    ret = -1;
+    goto cleanup;
+  }
+
+  snprintf(tar_cmd, 256, "tar -xf %s -C %s", PROC_KHEADERS_PATH, dirpath_tmp);
+  ret = system(tar_cmd);
+  if (ret) {
+    system(("rm -rf " + std::string(dirpath_tmp)).c_str());
+    goto cleanup;
+  }
+
+  /*
+   * If the new directory exists, it could have raced with a parallel
+   * extraction, in this case just delete the old directory and ignore.
+   */
+  ret = rename(dirpath_tmp, dirpath.c_str());
+  if (ret)
```

Comment:
if ret is ENOTEMPTY or EEXIST then we shouldn't return error from extract_kheaders, because kernel headers were actually successfully extracted, right? Otherwise loader won't use those headers in loader.cc:137 below.

Reply:
No. ret is overwritten by the following rm, see below

Reply:
Ah, you are right, sorry, I missed that.